Restore-CouponCode-After-Order-Cancellation

agorbulin opened the pull request:
restore coupon code usage after order cancellation

### Description (#19545)
**Problem:** when customer places order using personal (one time) coupon code
and order will be cancelled by PSP or manually, customer can not use it again.
**Magento ver.:** Reproduced on 2.2.* and 2.3.*

### Fixed Issues (#19545)
Restore CouponCode After Order Cancellation #19545

### Manual testing scenarios (*)
**STR:**
1. Create cart rule with coupon code (e.g. '1euro')
2. Set for rule 'Uses per Customer' = 1
3. Login as customer
4. Place product to the cart and apply coupon code (applied successfully)
5. Place order
6. Cancel order (manually in BO or by a PSP)
7. Place some product to the card again and apply coupon code one more time

**Expected result:**
Coupon code applied

**Actual result:**
Coupon code is not applied (error like this: Discount code "1euro" is not valid.)

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Restore-CouponCode-After-Order-Cancellation
restore couon usage after order cancellation

magento-engcom-team commented:
Hi @agorbulin. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

orlangur commented:
@agorbulin please report an issue separately so that such business logic change may be evaluated by PO.

agorbulin commented:
> @agorbulin please report an issue separately so that such business logic change may be evaluated by PO.
@orlangur 
Issue created #19545

Commit:
Restore-CouponCode-After-Order-Cancellation
take into account rule to_date, before restoring coupon code

Commit:
Merge branch '2.3-develop' into Restore-CouponCode-After-Order-Cancellation

agorbulin commented:
Verification of rule's to_date has been added, please check.

torhoehn commented on app/code/Magento/SalesRule/Observer/SalesOrderAfterCancelObserver.php:
```diff
@@ -0,0 +1,158 @@
+ *
+ * @package Magento\SalesRule\Observer
+ */
+class SalesOrderAfterCancelObserver implements ObserverInterface
```
Please cover new logic with unit test.

agorbulin replied:
@torhoehn 
Hi, I've added test, please review my commit below.

torhoehn reviewed (changes requested):
@agorbulin Thank you for the contribution. Please see the review notes.

Commit:
Magento2.3#19545 Restore CouponCode After Order Cancellation
- refactoring
- create unit test

magento-engcom-team commented:
Hi @torhoehn, thank you for the review.
[ENGCOM-3728](https://jira.corp.magento.com/browse/ENGCOM-3728) has been created to process this Pull Request

nmalevanec committed:
Merge remote-tracking branch 'mainline/2.3-develop' into Restore-CouponCode-After-Order-Cancellation

nmalevanec committed:
Fix static tests.

sidolov commented on app/code/Magento/SalesRule/Model/ResourceModel/Coupon/Usage.php:
```diff
@@ -25,11 +25,13 @@ protected function _construct()
+     * @throws \Magento\Framework\Exception\LocalizedException
      */
-    public function updateCustomerCouponTimesUsed($customerId, $couponId)
+    public function updateCustomerCouponTimesUsed($customerId, $couponId, $delta = 1)
```
Unfortunately, we can't change the public method signature according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

agorbulin replied:
Original method updateCustomerCouponTimesUsed restored
Custom method added instead
Please check.

Commit:
Restore-CouponCode-After-Order-Cancellation
- revert changes in original public method updateCustomerCouponTimesUsed
- add custom public method decrementCustomerCouponTimesUsed

Commit:
Merge branch '2.3-develop' into Restore-CouponCode-After-Order-Cancellation

magento-engcom-team commented:
Hi @sidolov, thank you for the review.
[ENGCOM-3728](https://jira.corp.magento.com/browse/ENGCOM-3728) has been created to process this Pull Request

sidolov commented:
Hi @agorbulin , we found the issue during QA activities: 
Exception error "Integrity constraint violation" when place order with applied coupon code
Manual testing scenario:

1. Create Cart Price Rule
2. Rule Name=test coupon
3. Websites=Main Website
4. Customer Groups=All
5. Coupon=Specific Coupon
6. Coupon Code=test
7. Uses per Customer=1
8. Apply=Percent of product price discount
9. Discount Amount=20
10. Login as customer
11. Add product to the cart and apply coupon code
12. Place order
13. Cancel order
14. Add product to the cart again and apply coupon code
15. Go to checkout/payment page
16. Click Place Order
17. Exception error is shown exception.log

```
[2019-01-29 09:41:27] main.CRITICAL: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '1-1' for key 'PRIMARY', query was: INSERT INTO `salesrule_coupon_usage` (`coupon_id`, `customer_id`, `times_used`) VALUES (?, ?, ?) {"exception":"[object] (Magento\\Framework\\DB\\Adapter\\DuplicateException(code: 1062): SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '1-1' for key 'PRIMARY', query was: INSERT INTO `salesrule_coupon_usage` (`coupon_id`, `customer_id`, `times_used`) VALUES (?, ?, ?) at /var/www/html/magento2ce/pr/lib/internal/Magento/Framework/DB/Adapter/Pdo/Mysql.php:588, Zend_Db_Statement_Exception(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '1-1' for key 'PRIMARY', query was: INSERT INTO `salesrule_coupon_usage` (`coupon_id`, `customer_id`, `times_used`) VALUES (?, ?, ?) at /var/www/html/magento2ce/pr/vendor/magento/zendframework1/library/Zend/Db/Statement/Pdo.php:235, PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '1-1' for key 'PRIMARY' at /var/www/html/magento2ce/pr/vendor/magento/zendframework1/library/Zend/Db/Statement/Pdo.php:228)"} []
```

sidolov commented:
@agorbulin , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!
